Add Notifications Panel Charter. Update Panels (#264)
solid/process: notifications-panel-charter.md

# Notifications Panel Charter

The mission of the Notifications Panel as part of the W3C Solid Community Group is to extend or define technical protocols and vocabularies to facilitate notification exchange on the Open Web Platform.

* Start date: 2021-09-13
* End date: 2022-12-31
* Charter extension: The charter extension history is documented in "About this charter".
* Confidentiality: Proceedings are Public
* Chair: Sarven Capadisli (Inrupt)
* Team Contact: Solid Editors
* Usual Meeting Schedule: Teleconferences: Weekly, although the chair may call for topic-specific calls in addition when needed and may change working mode as work progresses.

## Goals

> The Notifications Panel will create technical reports that standardise protocols and vocabularies for exchanging information as notifications. This should allow Web application developers to embed and facilitate access to social communication on the Web. The panel will explore use cases and requirements pertaining to activities in the Solid ecosystem; identify how existing specifications can be reused, including their specialisations and extensions; as well as develop new specifications. The work will be driven and refined by implementation experience. The panel will collaborate with other Solid panels and may take on or pass on challenges.

Reviewer: ```suggestion
The Notifications Panel will create technical reports that standardise protocols and vocabularies for exchanging information as notifications. This should allow Web application developers to embed and facilitate access to social communication on the Web. The panel will explore use cases and requirements pertaining to activities in the Solid ecosystem; identify how existing specifications, including their specialisations and extensions, can be reused; as well as develop new specifications. The work will be driven and refined by implementation experience. The panel will collaborate with other Solid panels and may take on or pass on challenges.
```

> The Notifications Panel will create technical reports that standardise protocols, vocabularies, and API for exchanging information as notifications. This should allow Web application developers to embed and facilitate access to social communication on the Web. The panel will explore use cases and requirements pertaining to activities in the Solid ecosystem, identify how existing specifications can be reused; including their specialisations and extensions, as well as develop new specifications. The work will be driven and refined by implementation experience. The panel will collaborate with other Solid panels and so may take on or pass on challenges.

Reviewer: ```suggestion
The Notifications Panel will create technical reports that standardise protocols, vocabularies, and APIs for exchanging information as notifications. This should allow Web application developers to embed and facilitate access to social communication on the Web. The panel will explore use cases and requirements pertaining to activities in the Solid ecosystem; identify how existing specifications can be reused, including their specialisations and extensions; as well as develop new specifications. The work will be driven and refined by implementation experience. The panel will collaborate with other Solid panels and may take on or pass on challenges.
```

> The Notifications Panel will create technical reports that standardise protocols, vocabularies, and API for exchanging information as status updates. This should allow Web application developers to embed and facilitate access to social communication on the Web. The panel will explore use cases and requirements pertaining to activities in the Solid ecosystem, identify how existing specifications can be reused; including their specialisations and extensions, as well as develop new specifications. The work will be driven and refined by implementation experience. The panel will collaborate with other Solid panels and so may take on or pass on challenges.

Reviewer 2: - What is a status update in this context?
- Does `social communication on the web` limit the scope of what notifications may or may not be related to?

Author: * Commited s/status update/notifications.
* No, I didn't mean to set any restrictions on what a notification may be. Web is all social at the end of the day. That's all.

> There are a number of use cases that the work of this panel will enable, including but not limited to:

Reviewer 2: Would seem more appropriate to have any/all use cases go through one UCR process?

Author: There was no intention to create multiple UCRs. What's listed is just some areas. Yes, a single UCR will be sufficient.

* User control of personal data: Some users would like to have autonomous control over their own social data, and share their data selectively across various systems.
* Activity subscriptions: Actors in the system may want information sent to them about particular shapes or when certain events or activities take place.
* Resource Access: Access to resources or services to request permission (in order to perform operations).

> * Service Actions: Such as activate services, create resources, create/update/delete user accounts, pod migration, pod archiving, delegating agents to perform tasks.

Reviewer: ```suggestion
* Service Actions: Such as activate services, create resources, create/update/delete user accounts, migrate pods, archive pods, or delegate agents to perform tasks.
```

> ## Scope

Reviewer 2: Generally looks good but should be informed by UCRs. How does acceptance of this charter influence the UCR process? For example, if a use case is raised after this charter is merged that is valid and legitimate, would this scope section be used as a reason to call it out of scope, or would this scope section be changed?

Author: It wasn't meant to be a strict list. Updated. We can certainly update the scope if need / agreed to.

> The panel within the framework of the W3C Solid CG and the Solid Project will determine the use cases that derive the requirements for the deliverables. Features that are not implemented due to time constraints can be put in a non-normative "roadmap" document for future work. The scope will include:

Reviewer: ```suggestion
The panel within the framework of the W3C Solid CG and the Solid Project will determine the use cases that derive the requirements for the deliverables. Features that are not implemented due to time constraints can be put into a non-normative "roadmap" document for future work. The scope will include:
```

* A transfer syntax for data such as activities should include at least the ability to describe the data using URIs in an extensible manner, time-stamping, and should include a serialization compatible with the RDF language.

* A protocol for exchanging social data should include at least the ability to share notifications using the transfer syntax developed by the Notifications Panel or extend syntaxes such as Activity Streams 2.0. This protocol may allow the capture of new data, the verification of data using techniques such as as digital signatures, server to server interactions, and the use of groups with some form of access control or capabilities.

> * Documentation detailing upgrade paths for existing technical reports or notes, including, but not limited to deprecated insecure WebSockets to secured upgrade of WebSockets.

Reviewer: ```suggestion
* Documentation detailing upgrade paths for existing technical reports or notes, including, but not limited to, upgrading deprecated insecure WebSockets to secured WebSockets.
```

> Other components necessary for building federated/decentralized social Web systems are in scope but will not lead to a recommendation without re-chartering, and should be discussed in the Notifications Panel.

Reviewer 3: What does this mean? 
'in scope' but 'will not lead to a recommendation' ?

Author: At this moment it is not clear if/what and to what detail we need to document upgrade paths. The only thing that's being roughly discussed is WebSockets but there can be other works. So we acknowledge that there are some works that's relevant to the CG/panel we can consider putting effort into - within "reason" - but what comes out of it may not be a normative document, like a spec. It could just be "Notes" / guidance to developers.

### Success Criteria

> In order to advance to technical report, the specification is expected to have two independent open source implementations of each feature defined in the specification. Independent implementations are developed by different parties and cannot share, reuse, or be derived from code of another qualifying implementation which is directly pertinent to the implementation of this specification.

Reviewer 3: Does this mean it cannot proceed if there are not two open source implementations?
What if there is one open and one closed?
For example ESS will have a closed source server implementation with open source client libraries. 
CSS will likely have an open source server implementation.
Then a server like Trinpod could have another closed source implementation.
Does this mean there would still not be enough implementations to advance to technical report?

Author: In order to publish a ~CR, the visibility of code plays a role in verifying and demonstrating adequate implementation experience of the work items towards an open standard.

I've pushed a commit replacing "open source" with "interoperable". While they have completely different meaning, it relaxes the requirement. Two implementations is a very low bar, and I would advise strongly against changing that. I'd propose even increasing that number especially if the CG is concerned about ensuring open source implementation experience for an open standard.

A closed source implementation could use open source code, and as far as I know, there is no reasonable way for the CG/public to catch that. Hence, it doesn't help towards showing maturity or wide use.

The specs can be informed by all (open and closed source) implementations nevertheless. "open source" is intended to keep one foot on... solid ground.

A full-blown Solid server is not required, or even a Solid server per se. What's assessed is whether "each feature defined in the specification" is implemented. What's not implemented gets reviewed by the panel e.g., to a consider whether an unimplemented feature should be removed from the spec or not.

## Deliverables

### Normative Deliverables

The panel will deliver the following to fulfill its goals, subject to discussion in the Solid CG:

* Notification Protocol

> * Notification Syntax

Reviewer 3: Should it specify that the specs need to contain rdf to enable test cases to link to the spec?

Author: Good question. I think that level of detail need not be in the charter, but if people would like to, happy to add that in. I was planning on making very loud noises about it any way :) We have over arching issues such as https://github.com/solid/specification/issues/6 ..

> * Notification Syntax

Reviewer 4: Do we look at defining Syntax or more of a Data Model which uses existing syntax ?

Author: There is a bit more detail about the syntax under scope. Meant to use "syntax" generally.. covering data model, vocabularies..

> In order to advance to technical report, the specification is expected to have two independent implementations of each feature defined in the specification. Independent implementations are developed by different parties and cannot share, reuse or derive form another qualifying implementation code which is directly pertinent to the implementation of this specification.

Reviewer: ```suggestion
In order to advance to technical report, the specification is expected to have two independent implementations of each feature defined in the specification. Independent implementations are developed by different parties and cannot share, reuse, or be derived from code of another qualifying implementation which is directly pertinent to the implementation of this specification.
```

* Notification Protocol (Discovery and Negotiation)
* Notification Data Model

> * Notification API

Reviewer 5: How is "Notification Protocol" being distinguished from "Notification API"? I would think that this group would define a "Notification Protocol" while a conforming implementation might deliver a "Notification API".

### Other Deliverables

The panel will deliver other non-normative documents, such as the following:

* A Use Cases and Requirements document defining how features in each specification relate to concrete use-cases.
* Test Suites for normative deliverables.
* Upgrade Notes

> May deliver:

Reviewer: ```suggestion
The panel may also deliver the following:
```

> Other non-normative documents may be delivered as:

Reviewer: ```suggestion
Other non-normative documents may be delivered, such as the following:
```

> * A use case document defining how features in each specification relate to concrete use-case.

Reviewer: ```suggestion
* A use case document defining how features in each specification relate to concrete use-cases.
```

* Primer or Best Practice documents to support Web developers when designing Solid applications.

### Milestones

The production of the deliverables depends upon the resources available, and will change as new information and implementation experience is reported to the group. The most up-to-date timeline is available from the Solid CG page.

> | Specification | ~FPWD   | ~LC     | ~ CR    | ~PR     | ~Rec    |

Reviewer: ```suggestion
| Specification | ~FPWD   | ~LC     | ~CR     | ~PR     | ~Rec    |
```

|---------------|---------|---------|---------|---------|---------|
| x             | 2021-Q4 | 2022-Q1 | 2022-Q2 | 2022-Q3 | 2022-Q4 |

Note: The Notifications Panel as part of the W3C Solid CG does not publish technical reports under the W3C Recommendation track. Thus, the milestones in the table above that are prefixed with "~" only communicate rough equivalence in maturity level of the technical reports according to section 6.2.1 of the W3C Process.

## Coordination

Technical coordination with the following Groups may be required:

### W3C Groups

* [Consent CG](https://www.w3.org/groups/cg/consent)
* [Credentials CG](https://www.w3.org/groups/cg/credentials)
* [JSON-LD WG](https://www.w3.org/groups/wg/json-ld)
* [Privacy CG](https://www.w3.org/groups/cg/privacycg)
* [RDF-DEV CG](https://www.w3.org/groups/cg/rdf-dev)
* [Social Web Incubator CG](https://www.w3.org/groups/cg/socialcg)
* [Web Applications WG](https://www.w3.org/groups/wg/webapps)

> Furthermore, the Notifications Panel expects to follow the following W3C Recommendations, Guidelines and Notes and, if necessary, to liaise with the communities behind the following documents:

Reviewer: ```suggestion
Furthermore, the Notifications Panel expects to follow the following W3C Recommendations, Guidelines, and Notes and, if necessary, to liaise with the communities behind them:
```

* Architecture of the World Wide Web, Volume I
* Internationalization Technical Reports and Notes

> * QA Framework: Specification Guidelines

Reviewer: Should link to all these groups & docs

### External Groups

* TBD

## Participation

Membership in the W3C Solid CG is required to participate in the Notifications Panel. All work and communication within the Solid CG is covered by the [Solid Code of Conduct](https://github.com/solid/process/blob/main/code-of-conduct.md) as well as the [Positive Work Environment at W3C: Code of Ethics and Professional Conduct](https://www.w3.org/Consortium/cepc/).

> To be successful, the Notifications Panel is expected to have 10 or more active participants for its duration and to have the participation of industry leaders in fields relevant to the specifications it produces. The Chairs and specification Editors are expected to contribute one to two days per week towards the Panel. The Chair may call occasional meetings consistent with the W3C Process requirements for meetings. There is no minimum requirement for other Participants.

Reviewer 3: What is the expectation of 10 or more participants is not met? 
If this paragraph has no consequence then how is it relevant?
I agree the chair and editors need to contribute 1 to 2 days per week. That means  we need to properly assess the time available for all that are nominated as chair?

Author: Say 1 chair, 4 editors/authors, 1 test suiter, 4 implementers = 10.

I have no issue to remove this paragraph. It is there as an expectation for its success, but you're right about consequences. If there isn't sufficient activity / contributions, we don't meet our goal. There are dips in the duration of the panel. Different contributors are more or less engaged at different times.

Noting that solid/process currently says at least 5 "members" - not even "participants" or "contributors"!  But again, that number can be anything. As per process, we could have 5 or 500 people as members and zero can show up. It wouldn't matter if the charter said 1, 5, or 500. The point is if there aren't *x* number of people engaging in the panel, there is no need to have a panel.

The Notifications Panel will also allocate the necessary resources for building test suites for each specification.

The group encourages questions and comments on its project repository, as described in Communication.

## Communication

Most Notifications Panel Teleconferences will be conducted on an as-needed basis. Normally, at least one teleconference will be held per week.

> Most of the technical work of the panel will be done through:

Reviewer 3: Should the charter also describe the weekly session on managing progress?

Author: I think that's also low level detail for a charter. Weekly sessions are while common, it is not required. It could be monthly or whenever there are enough number of participants interested in the agenda - people need to RSVP. We can use GitHub to help us manage/track progress.

* Repository: https://github.com/solid/notifications-panel
* Discussions: https://gitter.im/solid/notifications-panel

Information about the panel (for example, details about deliverables, issues, actions, status, and participants) will be available from the Notifications Panel repository.

## Decision Policy

> As explained in the W3C Process Document (section 3.3), this group will seek to make decisions when there is consensus and with due process. The expectation is that typically, an Editor or other participant makes an initial proposal, which is then refined in discussion with members of the group and other reviewers, and consensus emerges with little formal voting being required. However, if a decision is necessary for timely progress, but consensus is not achieved after careful consideration of the range of views presented, the Chairs should put a question out for voting within the group (allowing for remote asynchronous participation -- using, for example, chat channel or panel repository) and record a decision, along with any objections. The matter should then be considered resolved unless and until new information becomes available.

Reviewer 3: "...consensus emerges with little formal voting being required". Is there a time limit to this? How do you prevent it going on indefinitely or for an unreasonable amount of time?

".....for voting within the group and record a decision". Do you need to specify how that vote works e.g. majority? does chair have veto? does chair decide if its 50/50? etc. In other words making sure its clear and decisions can be made.

Author: First: voting is typically used to either re-check where we are on consensus (=straw poll) when new information is presented or as a last resort (=majority vote) when we can't reach (rough) consensus by other means e.g., objections are recorded and discussed, and if the issue is still unresolved (=commenter unsatisfied) in order to progress. The time limit is not fixed but we'll aim to resolve such issues within one meeting when all required participants are present. If still unresolved or there is an objection from a participant in a follow-up meeting when we are approving the prior meeting's minutes (in which the discussion or voting took place), they can present their case. Progress can't be blocked. Yes, the limit is determined by chair at some point in that they'll assess the situation, keep discussion DRY and focused, facilitate reaching consensus. If 50/50 (or even 51/49), then there definitely needs more discussion, demonstration of (open) implementation experience, using own tools, consider alternative proposals with less or weakest objections, implementation complexity, weigh participants' affiliations, break down the issue into smaller parts and re-vote to assess consensus... There is no one-size fits all factor when to wrap-up or what a representative majority may be. Depends on the issue and significance within a reasonable amount of time.

Reviewer 6: The Solid Protocol has included WebSocket-based live update for a long time.  It is core to the functionality of the platform.  Therefore, the solid spec  must at all times define how the live update works.

Now a change from "Web Sockets" to "Secure Web Sockets" version has been forced by the security problems with the original protocol -- that the connection was not authenticated.  There was agreement to move forward rapidly with this change.   Here are some ways of going this. 
1.  Abandon the insecure version and immediately upgrade all code to the secure one , with a security-scramble urgency to get the new code (which involves adding a token somewhere in the web socket data or the URI of the web socket).   Pick the first design a small group comes up with which looks good. Update NSS and CSS and ESS to the new spec. Add it to the test suite, add it to the spec. Release the spec Solid 1.0.  Allow future  notification architectures to evolve in good time, and later move to them in later releases.
2. Preserve the insecure version in the spec, and in NSS and CSS and ESS until the pane has come up with a well-thought out new design.  Immediately release Solid 0.9 with insecure web sockets.   Allow the panel time come up with a generic architecture and release Solid 1.0 with secure web sockets in due course.

I thought we were doing  1.   That would mean that if we use the interop panel it should be authorized to work on the security fix only over the next few weeks and not work on any more generic deigns until that secure fix is out and Solid 1.0 contains it  
(The Solid version numbers  here used to illustrate these alternatives are a bit arbitrary as we do not have a released series of them yet)

Author: >I thought we were doing 1.

Right, that is the consensus reached by the Editors team, and an issue that we agreed to follow-up on:  https://github.com/solid/specification/issues/289 .

It had no dependency on the notifications-panel.

I've proposed to re-open the panel because of other initiatives bubbling up in parallel and emphasised numerous times that it should not interfere with the priority and progress on Secure Web Sockets as we've originally agreed. As mentioned in https://github.com/solid/process/pull/264#issuecomment-911520108 , if contributors want to resolve that through the notifications-panel (because they can't without it), so be it. The key thing being whether that's a "few weeks" or possibly end of 2021-Q4 (as per notifications-panel Charter) but as part of a broader design based on informal feedback.
